components/common/src/command/package/install.rs

```diff
@@ -502,43 +625,139 @@ impl<'a> InstallTask<'a> {
 
     /// Adapter function to retrieve an installed package given an
     /// identifier, if it exists.
-    fn installed_package(&self, ident: &PackageIdent) -> Option<PackageInstall> {
-        PackageInstall::load(ident, Some(self.fs_root_path)).ok()
+    fn installed_package(&self, ident: &FullyQualifiedPackageIdent) -> Option<PackageInstall> {
+        PackageInstall::load(ident.as_ref(), Some(self.fs_root_path)).ok()
+    }
+
+    /// Checks for the latest installed package or cached artifact that matches a given package
+    /// identifier and returns a fully qualified package identifer if a match exists.
+    fn latest_installed_or_cached(
+        &self,
+        ident: &PackageIdent,
+    ) -> Result<FullyQualifiedPackageIdent> {
+        let latest_installed = self.latest_installed_ident(&ident);
+        let latest_cached = self.latest_cached_ident(&ident);
+        debug!(
+                "latest installed: {:?}, latest_cached: {:?}",
+                &latest_installed,
+                &latest_cached,
+            );
+        let latest = match (latest_installed, latest_cached) {
+            (Ok(pkg_install), Err(_)) => pkg_install,
+            (Err(_), Ok(pkg_artifact)) => pkg_artifact,
+            (Ok(pkg_install), Ok(pkg_artifact)) => {
+                if pkg_install.as_ref() > pkg_artifact.as_ref() {
+                    pkg_install
+                } else {
+                    pkg_artifact
+                }
+            }
+            (Err(_), Err(_)) => return Err(Error::PackageNotFound),
+        };
+        debug!("offline mode: winner: {:?}", &latest);
+
+        Ok(latest)
     }
 
-    // TODO (CM): This could return a plain bool IF we could ensure
-    // above that the package identifier is FULLY QUALIFIED
-    fn is_artifact_cached(&self, ident: &PackageIdent) -> Result<bool> {
-        Ok(self.cached_artifact_path(ident)?.is_file())
+    fn latest_installed_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        match PackageInstall::load(ident, Some(self.fs_root_path)) {
+            Ok(pi) => FullyQualifiedPackageIdent::from(pi.ident().clone()),
+            Err(_) => Err(Error::PackageNotFound),
+        }
+    }
+
+    fn latest_cached_ident(&self, ident: &PackageIdent) -> Result<FullyQualifiedPackageIdent> {
+        let filename_glob = {
+            let mut ident = ident.clone();
+            if ident.version.is_none() {
+                ident.version = Some(String::from("_VERSION_"));
+            }
+            if ident.release.is_none() {
+                ident.release = Some(String::from("_RELEASE_"));
+            }
+            match ident.archive_name() {
+                Some(name) => name.replace("_VERSION_", "?*").replace("_RELEASE_", "?*"),
+                None => return Err(Error::PackageNotFound),
+            }
```

Comment:
I must be missing something. Why wouldn't this be
```rust
            if ident.version.is_none() {
                ident.version = Some(String::from("?*"));
            }
            if ident.release.is_none() {
                ident.release = Some(String::from("?*"));
            }

            match ident.archive_name() {
                Some(name) => name
                None => return Err(Error::PackageNotFound),
            }
```
?

Reply:
Oh, this one! Yes maybe I should update that--I was being needlessly careful. When I saw the lack of validation present in `PackageIdent` I got nervous that at some future time we might want to apply rules to the `version` and `release` fields in a package identifier--like a release string should only have a set number of numeric characters. It was with this thought in mind that I put what I though would be a safer token to be replaced in the struct and then replaced over it once I had a `String` to work with.

I think I'll update to your suggestion and leave a note that if any future compile failures or issues arise it might be from stuffing arbitrary strings into these fields.

Reply:
I'm mildly nervous about stuffing regex strings into something that should be a version or a release. Could we not just assemble the regex string directly?